syntax = "proto3";

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

message FrameMessage {
    int32 frame_number = 1;
    google.protobuf.Timestamp frame_time = 2;
    bytes image_data = 3;
}

service FrameService {
    rpc SendFrame(FrameMessage) returns (google.protobuf.Empty) {}
}

////////////============== Configuration Service ============/////////////

message CameraSettings {
    int32 threshold = 1;
    int32 max_diff_pixels = 2;
}

message GrpcSettings {
    string server_ip_address = 1;
    string camera_ip_address = 2;
    string port_number = 3;
}

message YoloSettings {
    string class_list_path_windows = 1;
    string class_list_path_linux = 2;
    string net_path_windows = 3;
    string net_path_linux = 4;
    int32 input_width = 5;
    int32 input_height = 6;
    float score_threshold = 7;
    float nms_threshold = 8;
    float confidence_threshold = 9;
}

message LogSettings {
    string log_level = 1;
    string log_directory = 2;
}

message OutputSettings {
    string video_path_windows = 1;
    string video_path_linux = 2;
    string db_path_windows = 3;
    string db_path_linux = 4;
    string image_path_windows = 5;
    string image_path_linux = 6;
}

message Configuration {
    CameraSettings camera_settings = 1;
    GrpcSettings grpc_settings = 2;
    YoloSettings yolo_settings = 3;
    LogSettings log_settings = 4;
    OutputSettings output_settings = 5;
}
service ConfigurationService {
    rpc UpdateCameraSettings(CameraSettings) returns (google.protobuf.Empty) {}
    rpc UpdateGrpcSettings(GrpcSettings) returns (google.protobuf.Empty) {}
    rpc UpdateYoloSettings(YoloSettings) returns (google.protobuf.Empty) {}
    rpc UpdateLogSettings(LogSettings) returns (google.protobuf.Empty) {}
    rpc UpdateOutputSettings(OutputSettings) returns (google.protobuf.Empty) {}
}